src/dr_llm/workers/worker_controller.py

```
@@ -1,72 +1,80 @@
+        if self._final_snapshot is not None:
+            return self._final_snapshot
+
+        with self._join_lock:
+            if self._final_snapshot is not None:
+                return self._final_snapshot
+
+            _, not_done = wait(self._futures, timeout=timeout)
+            all_done = not not_done
+            try:
+                if not_done:
+                    raise TimeoutError("Timed out waiting for workers to stop")
```

Ruff (0.15.9)

[warning] Avoid specifying long messages outside the exception class

(TRY003)